### DISCUSSION TITLE
[WS2122] Homework 7 Discussion

### DISCUSSION THREAD

**Discussion by changkun** [mimuc/gp-ws2021#72] [Discussion] [answered]
This thread is open for the discussion of homework 7.

Please also note that the deadline for the 7th homework project is 07.02.2022 23:59:59 (CET)

**Comment by Consti10** [comment, marked as answer]
Hello,
I have a couple of questions regarding this project:
1) The missing GPU support really is a limiting factor. I am working on Ubuntu 20.04 on a Laptop with GTX 1070 Max-Q GPU and the proprietary nvida drivers installed. However, I cannot even install NVIDIA CUDA, let alone install pytorch3D wth CUDA enabled.
My unsuccessfull quest on getting these 2 to work included messing up my ubuntu installation with incompatible nvidia drivers, so I think I have to give up on pytorch3d with GPU support on my laptop.

Therefore, my first question is - can we get an environment (aka remote) with some working NVDIA GPU where we can install CUDA,CUBE, pytorch and pytorch3D with CUDA enabled, or is this even worth the effort (what is the speedup one can expect from CUDA on this project)

2) Since I am working with CPU only right now, I started by using the "chamfer_distance" instead of a MSE on the rendered images,
and greatly reduced the n of uniform samples taken from the mesh per iteration (similar to the dolphin examle from the tutorial).
One can already observe some interesting things here - for example, while the dolphin from the tutorial "looks fine" after a couple 1000s of iterations, the bunny has some problems.
Here I was using:
N_SAMPLE_POINTS= 10000 (how many verts are used for the chamfer_distance step in each optimzing iteration
and N_ITER= 1000 (N of optimization iterations)
All this takes ~15min on my CPU:


![Screenshot from 2022-01-26 00-35-05](https://user-images.githubusercontent.com/16462285/151077804-434bf9c7-7edd-40dd-85eb-f895d1d773a2.png)

![Screenshot from 2022-01-26 00-37-59](https://user-images.githubusercontent.com/16462285/151077855-078bcdeb-dcce-4a33-98d4-6d5f3feb8418.png)

![Screenshot from 2022-01-26 00-34-17](https://user-images.githubusercontent.com/16462285/151077869-a6098f70-66bf-4914-b0ff-8c76c60a5f74.png)


While I understand the project such that we shall not use "chamfer_distance" but rather the MSE of the rendered images, If I add this code to to my optimizing loop, the added execution time of rendering the image in each optimization loops makes this task impossible to complete on a CPU.

**Reply by changkun** [reply]
Regarding the GPU support, we can support a clean machine without any installation, but you will need to install CUDA/PyTorch3D yourself. As I do not understand why CUDA can't be installed in your environment, I'd still recommend trying to figure out the reason and install them.

Regarding the choice of losses, perhaps the comment is a bit distracting from thinking of a new solution. In general, the losses are not limited to chamfer but also can be a combination of multiple ones. Depending on the reasoning, one could design other losses for different tasks. Theoretically, a loss does not influence the model capacity but only may constraints you from searching for a good model using an optimizer.

The comment might also be distracting and misleading regarding the expected iterations, as a good solution might be more efficient and fewer iterations required. For 2000 iterations on a 2080Ti, it only requires less than 1 minute. However, for some reason, might be a bug in PyTorch3D, later iterations are significantly slower than earlier ones. In my constructed solution for creating this project, it takes roughly 30 minutes. (I may rethink a more efficient one in order to have it be presented in the class)
